src/V3Assert.cpp

```diff
@@ -353,10 +360,40 @@ class AssertVisitor final : public VNVisitor {
         }
         nodep->replaceWith(inp);
     }
+
+    //========== Move $sampled down to read-only variables
     virtual void visit(AstSampled* nodep) override {
+        if (nodep->user1()) return;
+        VL_RESTORER(m_inSampled);
+        {
+            m_inSampled = true;
+            iterateChildren(nodep);
+        }
         nodep->replaceWith(nodep->exprp()->unlinkFrBack());
         VL_DO_DANGLING(pushDeletep(nodep), nodep);
     }
+    virtual void visit(AstVarRef* nodep) override {
+        iterateChildren(nodep);
+        if (!m_inSampled) return;
+        if (!nodep->access().isReadOnly()) {
+            nodep->v3warn(E_UNSUPPORTED,
+                          "Unsupported: Assignment to variable inside sampled expression");
```

Comment:
Let's reword this a bit.
```suggestion
                          "Unsupported: Write to variable in sampled expression");
```

Reply:
Done.

Reply:
Update the `*.out` file as well.

Reply:
I've pushed it meanwhile.